File: Olympiad/COCI/official/2007/regional_solutions/tetris.pas

(*
  Croatian Regional Competition in Informatics 2007
  Task TETRIS
  Programming language Pascal
*)

program tetris;

const MAXS = 100;

var
   v           : array[1..MAXS] of longint;
   s, f, i, rj : longint;

begin
   readln(s, f);
   if (f=4)or(f=7) then
   begin
      dec(f);
      for i:=1 to s do read(v[s-i+1]);
   end else
      for i:=1 to s do read(v[i]);

   rj := 0;
   case f of
     1:  begin
            rj:=s;
            for i:=1 to s-3 do
               if (v[i]=v[i+1])and(v[i]=v[i+2])and(v[i]=v[i+3]) then inc(rj);
         end;
     2:  begin
            for i:=1 to s-1 do
               if v[i]=v[i+1] then inc(rj);
         end;
     3:  begin
            for i:=1 to s-1 do
               if v[i]=v[i+1]+1 then inc(rj);
            for i:=1 to s-2 do
               if (v[i]=v[i+1])and(v[i]=v[i+2]-1) then inc(rj);
         end;
     5:  begin
            for i:=1 to s-2 do
               if ((v[i]-v[i+1]=1)or(v[i]-v[i+1]=0))and(v[i]=v[i+2]) then inc(rj);
            for i:=1 to s-1 do
               if abs(v[i]-v[i+1])=1 then inc(rj);
         end;
     6:  begin
            for i:=1 to s-2 do
               if (v[i+1]=v[i+2])and((v[i+1]-v[i]=1)or(v[i+1]-v[i]=0)) then inc(rj);
            for i:=1 to s-1 do
               if (v[i]=v[i+1])or(v[i]=v[i+1]+2) then inc(rj);
         end;
   end;
   writeln(rj);
end.

File: Olympiad/COCI/official/2007/regional_solutions/tetris.cpp

/*
  Croatian Regional Competition in Informatics 2007
  Task TETRIS
  Programming language C++
*/

#include <cstdio>
#include <cstring>

using namespace std;

int S, T;
int h[100];

int check( char *a ) {
   int n = strlen( a );

   int ret = 0;
   for( int i = 0; i+n <= S; ++i ) {
      int ok = 1;
      for( int j = 0; j < n; ++j )
         if( a[j] - h[i+j] != a[0] - h[i] )
            ok = 0;
      if( ok ) ++ret;
   }
   return ret;
}

int main( void ) {

   scanf( "%d%d", &S, &T );
   for( int i = 0; i < S; ++i ) scanf( "%d", &h[i] );

   int ret = 0;
   if( T == 1 ) ret = check( "0" ) + check( "0000" );
   if( T == 2 ) ret = check( "00" );
   if( T == 3 ) ret = check( "001" ) + check( "10" );
   if( T == 4 ) ret = check( "100" ) + check( "01" );
   if( T == 5 ) ret = check( "000" ) + check( "01" ) + check( "101" ) + check( "10" );
   if( T == 6 ) ret = check( "000" ) + check( "00" ) + check( "011" ) + check( "20" );
   if( T == 7 ) ret = check( "000" ) + check( "00" ) + check( "110" ) + check( "02" );

   printf( "%d\n", ret );

   return 0;
}
